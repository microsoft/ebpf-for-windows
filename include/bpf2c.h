// Copyright (c) Microsoft Corporation
// SPDX-License-Identifier: MIT

#pragma once
#if defined(NO_CRT)
typedef signed char int8_t;
typedef short int16_t;
typedef int int32_t;
typedef long long int64_t;
typedef unsigned char uint8_t;
typedef unsigned short uint16_t;
typedef unsigned int uint32_t;
typedef unsigned long long uint64_t;
#define bool _Bool
#define false 0
#define true 1
#define UINT32_MAX ((uint32_t)0xFFFFFFFF)

#else
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#endif

#include "ebpf_structs.h"

#ifdef __cplusplus
extern "C"
{
#endif

#define UBPF_STACK_SIZE 512

#define IMMEDIATE(X) (int32_t) X
#define OFFSET(X) (int16_t) X
#define POINTER(X) (uint64_t)(X)

#if !defined(htobe16)
#define htobe16(X) swap16(X)
#define htobe32(X) swap32(X)
#define htobe64(X) swap64(X)

#define htole16(X) (X)
#define htole32(X) (X)
#define htole64(X) (X)
#endif

    typedef struct _helper_function_entry
    {
        uint64_t (*address)(uint64_t r1, uint64_t r2, uint64_t r3, uint64_t r4, uint64_t r5);
        uint32_t helper_id;
        const char* name;
        bool tail_call;
    } helper_function_entry_t;

    typedef struct _map_entry
    {
        void* address;
        ebpf_map_definition_in_file_t definition;
        const char* name;
    } map_entry_t;

    typedef struct _program_entry
    {
        // DLLs put the strings into the same section, so add a marker
        // at the start of a program entry to make it easy to find
        // entries in the programs section.
        uint64_t zero;

        uint64_t (*function)(void*);
        const char* pe_section_name;
        const char* section_name;
        const char* program_name;
        uint16_t* referenced_map_indices;
        uint16_t referenced_map_count;
        helper_function_entry_t* helpers;
        uint16_t helper_count;
        size_t bpf_instruction_count;
        ebpf_program_type_t* program_type;
        ebpf_attach_type_t* expected_attach_type;
        const uint8_t* program_info_hash;
        size_t program_info_hash_length;
    } program_entry_t;

    typedef struct _metadata_table
    {
        void (*programs)(_Outptr_result_buffer_maybenull_(*count) program_entry_t** programs, _Out_ size_t* count);
        void (*maps)(_Outptr_result_buffer_maybenull_(*count) map_entry_t** maps, _Out_ size_t* count);
        void (*hash)(_Outptr_result_buffer_maybenull_(*size) const uint8_t** hash, _Out_ size_t* size);
    } metadata_table_t;

    inline uint16_t
    swap16(uint16_t value)
    {
        return value << 8 | value >> 8;
    }

    inline uint32_t
    swap32(uint32_t value)
    {
        return swap16(value >> 16) | ((uint32_t)swap16(value & ((1 << 16) - 1))) << 16;
    }

    inline uint64_t
    swap64(uint64_t value)
    {
        return swap32(value >> 32) | ((uint64_t)swap32(value & ((1ull << 32ull) - 1))) << 32;
    }

    void
    division_by_zero(uint32_t address);

#ifdef __cplusplus
}
#endif
