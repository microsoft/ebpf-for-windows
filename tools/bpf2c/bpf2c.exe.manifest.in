<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<!--
  Copyright (c) Microsoft Corporation
  SPDX-License-Identifier: MIT

  Overrides the default heap implementation with segment heap.
-->
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0" xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
  <assemblyIdentity name="microsoft.ebpf-for-windows.bpf2c" type="win32" version="{version}.0"></assemblyIdentity>
  <asmv3:application>
    <asmv3:windowsSettings xmlns="http://schemas.microsoft.com/SMI/2020/WindowsSettings">
      <heapType>SegmentHeap</heapType>
    </asmv3:windowsSettings>
  </asmv3:application>
</assembly>
