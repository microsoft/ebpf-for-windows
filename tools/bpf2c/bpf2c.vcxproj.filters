<?xml version="1.0" encoding="utf-8"?>
<!--
  Copyright (c) Microsoft Corporation
  SPDX-License-Identifier: MIT
-->
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <Filter Include="Source Files">
      <UniqueIdentifier>{4FC737F1-C7A5-4376-A066-2A32D752A2FF}</UniqueIdentifier>
      <Extensions>cpp;c;cc;cxx;c++;cppm;ixx;def;odl;idl;hpj;bat;asm;asmx</Extensions>
    </Filter>
    <Filter Include="Header Files">
      <UniqueIdentifier>{93995380-89BD-4b04-88EB-625FBE52EBFB}</UniqueIdentifier>
      <Extensions>h;hh;hpp;hxx;h++;hm;inl;inc;ipp;xsd</Extensions>
    </Filter>
    <Filter Include="Resource Files">
      <UniqueIdentifier>{67DA6AB6-F800-4c08-8B7A-83BB121AAD01}</UniqueIdentifier>
      <Extensions>rc;ico;cur;bmp;dlg;rc2;rct;bin;rgs;gif;jpg;jpeg;jpe;resx;tiff;tif;png;wav;mfcribbon-ms</Extensions>
    </Filter>
  </ItemGroup>
  <ItemGroup>
    <ClCompile Include="bpf2c.cpp">
      <Filter>Source Files</Filter>
    </ClCompile>
    <ClCompile Include="bpf_code_generator.cpp">
      <Filter>Source Files</Filter>
    </ClCompile>
  </ItemGroup>
  <ItemGroup>
    <ClInclude Include="..\..\include\bpf2c.h">
      <Filter>Header Files</Filter>
    </ClInclude>
    <ClInclude Include="bpf_code_generator.h">
      <Filter>Header Files</Filter>
    </ClInclude>
    <ClInclude Include="..\..\external\ebpf-verifier\src\btf_parser.h">
      <Filter>Header Files</Filter>
    </ClInclude>
  </ItemGroup>
  <ItemGroup>
    <CustomBuild Include="bpf2c_dll.c">
      <Filter>Source Files</Filter>
    </CustomBuild>
    <CustomBuild Include="bpf2c_driver.c">
      <Filter>Source Files</Filter>
    </CustomBuild>
  </ItemGroup>
</Project>