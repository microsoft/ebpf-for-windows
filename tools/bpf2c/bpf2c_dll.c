// Copyright (c) Microsoft Corporation
// SPDX-License-Identifier: MIT

#include "bpf2c.h"

#include <stdio.h>
#define WIN32_LEAN_AND_MEAN // Exclude rarely-used stuff from Windows headers
#include <windows.h>

#define metadata_table ___METADATA_TABLE___##_metadata_table
extern metadata_table_t metadata_table;

BOOL APIENTRY
DllMain(_In_ HMODULE hModule, unsigned int ul_reason_for_call, _In_ void* lpReserved)
{
    UNREFERENCED_PARAMETER(hModule);
    UNREFERENCED_PARAMETER(lpReserved);
    switch (ul_reason_for_call) {
    case DLL_PROCESS_ATTACH:
    case DLL_THREAD_ATTACH:
    case DLL_THREAD_DETACH:
    case DLL_PROCESS_DETACH:
        break;
    }
    return TRUE;
}

__declspec(dllexport) metadata_table_t* get_metadata_table() { return &metadata_table; }
