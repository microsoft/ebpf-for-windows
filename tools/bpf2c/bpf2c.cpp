// Copyright (c) Microsoft Corporation
// SPDX-License-Identifier: MIT

#define _SILENCE_CXX17_CODECVT_HEADER_DEPRECATION_WARNING

#include <Windows.h>
#include <bcrypt.h>
#include <codecvt>
#include <functional>
#include <fstream>
#include <iostream>
#include <map>
#include <regex>
#include <sstream>
#include <string>
#include <tuple>
#include <vector>

#undef max
#include "bpf_code_generator.h"
#include "ebpf_api.h"

#pragma comment(lib, "Bcrypt.lib")

const char copyright_notice[] = "// Copyright (c) Microsoft Corporation\n// SPDX-License-Identifier: MIT\n";

const char bpf2c_driver[] =
#include "bpf2c_driver.template"
    ;

const char bpf2c_dll[] =
#include "bpf2c_dll.template"
    ;

class _hash
{
  public:
    _hash(const std::string& algorithm)
    {
        HRESULT hr;
        std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> converter;
        std::wstring wide_algorithm = converter.from_bytes(algorithm);

        hr = BCryptOpenAlgorithmProvider(&algorithm_handle, wide_algorithm.c_str(), nullptr, BCRYPT_HASH_REUSABLE_FLAG);
        if (!SUCCEEDED(hr)) {
            throw std::runtime_error(
                std::string("BCryptOpenAlgorithmProvider failed with algorithm: ") + algorithm +
                " HR=" + std::to_string(hr));
        }
    }
    ~_hash() { BCryptCloseAlgorithmProvider(algorithm_handle, 0); }

    std::vector<uint8_t>
    hash_string(const std::string& data)
    {
        uint32_t hash_length;
        std::vector<uint8_t> hash;
        BCRYPT_HASH_HANDLE hash_handle;
        HRESULT hr;
        hr = BCryptCreateHash(algorithm_handle, &hash_handle, nullptr, 0, nullptr, 0, 0);
        if (!SUCCEEDED(hr)) {
            throw std::runtime_error(std::string("BCryptCreateHash failed with HR=") + std::to_string(hr));
        }
        hr = BCryptHashData(
            hash_handle,
            reinterpret_cast<uint8_t*>(const_cast<char*>(data.data())),
            static_cast<unsigned long>(data.size()),
            0);
        if (!SUCCEEDED(hr)) {
            BCryptDestroyHash(hash_handle);
            throw std::runtime_error(std::string("BCryptHashData failed with HR=") + std::to_string(hr));
        }

        unsigned long bytes_written;
        hr = BCryptGetProperty(
            algorithm_handle,
            BCRYPT_HASH_LENGTH,
            reinterpret_cast<uint8_t*>(&hash_length),
            sizeof(hash_length),
            &bytes_written,
            0);
        if (!SUCCEEDED(hr)) {
            BCryptDestroyHash(hash_handle);
            throw std::runtime_error(
                std::string("BCryptGetProperty failed with BCRYPT_HASH_LENGTH  HR=") + std::to_string(hr));
        }
        hash.resize(hash_length);
        hr = BCryptFinishHash(hash_handle, hash.data(), static_cast<unsigned long>(hash.size()), 0);
        BCryptDestroyHash(hash_handle);
        if (!SUCCEEDED(hr)) {
            throw std::runtime_error(std::string("BCryptFinishHash failed with HR=") + std::to_string(hr));
        }
        return hash;
    }

  private:
    BCRYPT_ALG_HANDLE algorithm_handle;
};

void
emit_skeleton(const std::string& c_name, const std::string& code)
{
    auto output = std::regex_replace(code, std::regex(std::string("___METADATA_TABLE___")), c_name);
    output = output.substr(strlen(copyright_notice) + 1);
    std::cout << output << std::endl;
}

std::string
load_file_to_memory(const std::string& path)
{
    struct stat st;
    if (stat(path.c_str(), &st)) {
        throw std::runtime_error(std::string("Failed to read file: ") + path);
    }
    if (std::ifstream stream{path, std::ios::in | std::ios::binary}) {
        std::string data;
        data.resize(st.st_size);
        if (!stream.read(data.data(), data.size())) {
            throw std::runtime_error(std::string("Failed to read file: ") + path);
        }
        return data;
    }
    throw std::runtime_error(std::string("Failed to read file: ") + path);
}

int
main(int argc, char** argv)
{
    try {
        enum class output_type
        {
            Bare,
            KernelPE,
            UserPE,
        } type = output_type::Bare;
        std::string verifier_output_file;
        std::string file;
        std::string hash_algorithm = "SHA256";
        bool verify_programs = true;
        std::vector<std::string> parameters(argv + 1, argv + argc);
        auto iter = parameters.begin();
        auto iter_end = parameters.end();
        std::map<std::string, std::tuple<std::string, std::function<bool()>>> options = {
            {"--sys",
             {"Generate code for a Windows driver",
              [&]() {
                  type = output_type::KernelPE;
                  return true;
              }}},
            {"--dll",
             {"Generate code for a Windows DLL",
              [&]() {
                  type = output_type::UserPE;
                  return true;
              }}},
#if defined(ENABLE_SKIP_VERIFY)
            {"--no-verify",
             {"Skip validating code using verifier",
              [&]() {
                  verify_programs = false;
                  return true;
              }}},
#endif
            {"--bpf",
             {"Input ELF file containing BPF byte code",
              [&]() {
                  ++iter;
                  if (iter == iter_end) {
                      std::cerr << "Invalid --bpf option" << std::endl;
                      return false;
                  } else {
                      file = *iter;
                      return true;
                  }
              }}},
            {"--hash",
             {"Algorithm used to hash ELF file",
              [&]() {
                  ++iter;
                  if (iter == iter_end) {
                      std::cerr << "Invalid --hash option" << std::endl;
                      return false;
                  } else {
                      hash_algorithm = *iter;
                      return true;
                  }
              }}},
            {"--help",
             {"This help menu",
              [&]() {
                  std::cerr << argv[0]
                            << " is a tool to generate C code"
                               "from an ELF file containing BPF byte code."
                            << std::endl;
                  std::cerr << "Options are:" << std::endl;
                  for (auto [option, tuple] : options) {
                      auto [help, _] = tuple;
                      std::cerr << option.c_str() << "\t" << help.c_str() << std::endl;
                  }
                  return false;
              }}},
        };

        for (; iter != iter_end; ++iter) {
            auto option = options.find(*iter);
            if (option == options.end()) {
                option = options.find("--help");
            }
            auto [_, function] = option->second;
            if (!function()) {
                return 1;
            }
        }

        std::string c_name = file.substr(file.find_last_of("\\") + 1);
        c_name = c_name.substr(0, c_name.find("."));
        auto data = load_file_to_memory(file);
        std::optional<std::vector<uint8_t>> hash_value;
        if (hash_algorithm != "none") {

            _hash hash(hash_algorithm);
            hash_value = hash.hash_string(data);
        }
        auto stream = std::stringstream(data);
        // TODO: Issue #834 - validate the ELF is well formed
        bpf_code_generator generator(stream, c_name, {hash_value});

        // Capture list of sections.
        std::vector<bpf_code_generator::unsafe_string> sections = generator.program_sections();

        if (verify_programs && sections.size() == 0) {
            std::cerr << "ELF " << file << " file contains no program sections" << std::endl;
            return 1;
        }

        // Parse global data.
        generator.parse();

        // Parse per-section data.
        for (const auto& section : sections) {
            ebpf_program_type_t program_type;
            ebpf_attach_type_t attach_type;
            // TODO: Issue #1172
            // Workaround: If querying the program and attach type fails, default it to XDP until Issue #1172
            // is fixed.
            if (ebpf_get_program_type_by_name(section.raw().c_str(), &program_type, &attach_type) != EBPF_SUCCESS) {
                program_type = EBPF_PROGRAM_TYPE_XDP;
                attach_type = EBPF_ATTACH_TYPE_XDP;
            }
            const char* report = nullptr;
            const char* error_message = nullptr;
            ebpf_api_verifier_stats_t stats;
            if (verify_programs && ebpf_api_elf_verify_section_from_memory(
                                       data.c_str(),
                                       data.size(),
                                       section.raw().c_str(),
                                       &program_type,
                                       false,
                                       &report,
                                       &error_message,
                                       &stats) != 0) {
                report = ((report == nullptr) ? "" : report);
                throw std::runtime_error(
                    std::string("Verification failed for ") + section.raw() + std::string(" with error ") +
                    std::string(error_message) + std::string("\n Report:\n") + std::string(report));
            }
            generator.parse(section, program_type, attach_type);
        }

        for (const auto& section : sections) {
            generator.generate(section);
        }

        std::cout << copyright_notice << std::endl;
        std::cout << "// Do not alter this generated file." << std::endl;
        std::cout << "// This file was generated from " << file << std::endl << std::endl;
        switch (type) {
        case output_type::Bare:
            break;
        case output_type::KernelPE:
            emit_skeleton(c_name, bpf2c_driver);
            break;
        case output_type::UserPE:
            emit_skeleton(c_name, bpf2c_dll);
            break;
        }
        generator.emit_c_code(std::cout);
    } catch (std::runtime_error err) {
        std::cerr << err.what() << std::endl;
        return 1;
    }
    return 0;
}
