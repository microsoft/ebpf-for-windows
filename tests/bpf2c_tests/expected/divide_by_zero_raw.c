// Copyright (c) Microsoft Corporation
// SPDX-License-Identifier: MIT

// Do not alter this generated file.
// This file was generated from divide_by_zero.o

#include "bpf2c.h"

static void
_get_hash(_Outptr_result_buffer_maybenull_(*size) const uint8_t** hash, _Out_ size_t* size)
{
    *hash = NULL;
    *size = 0;
}
#pragma data_seg(push, "maps")
static map_entry_t _maps[] = {
    {NULL,
     {
         BPF_MAP_TYPE_ARRAY, // Type of map.
         4,                  // Size in bytes of a map key.
         4,                  // Size in bytes of a map value.
         1,                  // Maximum number of entries allowed in the map.
         0,                  // Inner map index.
         PIN_NONE,           // Pinning type for the map.
         10,                 // Identifier for a map template.
         0,                  // The id of the inner map template.
     },
     "test_map"},
};
#pragma data_seg(pop)

static void
_get_maps(_Outptr_result_buffer_maybenull_(*count) map_entry_t** maps, _Out_ size_t* count)
{
    *maps = _maps;
    *count = 1;
}

static helper_function_entry_t divide_by_zero_helpers[] = {
    {NULL, 1, "helper_id_1"},
};

static GUID divide_by_zero_program_type_guid = {
    0xf1832a85, 0x85d5, 0x45b0, {0x98, 0xa0, 0x70, 0x69, 0xd6, 0x30, 0x13, 0xb0}};
static GUID divide_by_zero_attach_type_guid = {
    0x85e0d8ef, 0x579e, 0x4931, {0xb0, 0x72, 0x8e, 0xe2, 0x26, 0xbb, 0x2e, 0x9d}};
static uint16_t divide_by_zero_maps[] = {
    0,
};

#pragma code_seg(push, "xdp")
static uint64_t
divide_by_zero(void* context)
#line 31 "sample/divide_by_zero.c"
{
#line 31 "sample/divide_by_zero.c"
    // Prologue
#line 31 "sample/divide_by_zero.c"
    uint64_t stack[(UBPF_STACK_SIZE + 7) / 8];
#line 31 "sample/divide_by_zero.c"
    register uint64_t r0 = 0;
#line 31 "sample/divide_by_zero.c"
    register uint64_t r1 = 0;
#line 31 "sample/divide_by_zero.c"
    register uint64_t r2 = 0;
#line 31 "sample/divide_by_zero.c"
    register uint64_t r3 = 0;
#line 31 "sample/divide_by_zero.c"
    register uint64_t r4 = 0;
#line 31 "sample/divide_by_zero.c"
    register uint64_t r5 = 0;
#line 31 "sample/divide_by_zero.c"
    register uint64_t r6 = 0;
#line 31 "sample/divide_by_zero.c"
    register uint64_t r10 = 0;

#line 31 "sample/divide_by_zero.c"
    r1 = (uintptr_t)context;
#line 31 "sample/divide_by_zero.c"
    r10 = (uintptr_t)((uint8_t*)stack + sizeof(stack));

    // EBPF_OP_MOV64_IMM pc=0 dst=r6 src=r0 offset=0 imm=0
#line 31 "sample/divide_by_zero.c"
    r6 = IMMEDIATE(0);
    // EBPF_OP_STXW pc=1 dst=r10 src=r6 offset=-4 imm=0
#line 33 "sample/divide_by_zero.c"
    *(uint32_t*)(uintptr_t)(r10 + OFFSET(-4)) = (uint32_t)r6;
    // EBPF_OP_MOV64_REG pc=2 dst=r2 src=r10 offset=0 imm=0
#line 33 "sample/divide_by_zero.c"
    r2 = r10;
    // EBPF_OP_ADD64_IMM pc=3 dst=r2 src=r0 offset=0 imm=-4
#line 33 "sample/divide_by_zero.c"
    r2 += IMMEDIATE(-4);
    // EBPF_OP_LDDW pc=4 dst=r1 src=r0 offset=0 imm=0
#line 34 "sample/divide_by_zero.c"
    r1 = POINTER(_maps[0].address);
    // EBPF_OP_CALL pc=6 dst=r0 src=r0 offset=0 imm=1
#line 34 "sample/divide_by_zero.c"
    r0 = divide_by_zero_helpers[0].address
#line 34 "sample/divide_by_zero.c"
         (r1, r2, r3, r4, r5);
#line 34 "sample/divide_by_zero.c"
    if ((divide_by_zero_helpers[0].tail_call) && (r0 == 0))
#line 34 "sample/divide_by_zero.c"
        return 0;
    // EBPF_OP_JEQ_IMM pc=7 dst=r0 src=r0 offset=3 imm=0
#line 35 "sample/divide_by_zero.c"
    if (r0 == IMMEDIATE(0))
#line 35 "sample/divide_by_zero.c"
        goto label_1;
    // EBPF_OP_LDXW pc=8 dst=r1 src=r0 offset=0 imm=0
#line 36 "sample/divide_by_zero.c"
    r1 = *(uint32_t*)(uintptr_t)(r0 + OFFSET(0));
    // EBPF_OP_MOV64_IMM pc=9 dst=r6 src=r0 offset=0 imm=100000
#line 36 "sample/divide_by_zero.c"
    r6 = IMMEDIATE(100000);
    // EBPF_OP_DIV64_REG pc=10 dst=r6 src=r1 offset=0 imm=0
#line 36 "sample/divide_by_zero.c"
    r6 = r1 ? (r6 / r1) : 0;
label_1:
    // EBPF_OP_MOV64_REG pc=11 dst=r0 src=r6 offset=0 imm=0
#line 39 "sample/divide_by_zero.c"
    r0 = r6;
    // EBPF_OP_EXIT pc=12 dst=r0 src=r0 offset=0 imm=0
#line 39 "sample/divide_by_zero.c"
    return r0;
#line 39 "sample/divide_by_zero.c"
}
#pragma code_seg(pop)
#line __LINE__ __FILE__

#pragma data_seg(push, "programs")
static program_entry_t _programs[] = {
    {
        0,
        divide_by_zero,
        "xdp",
        "xdp",
        "divide_by_zero",
        divide_by_zero_maps,
        1,
        divide_by_zero_helpers,
        1,
        13,
        &divide_by_zero_program_type_guid,
        &divide_by_zero_attach_type_guid,
    },
};
#pragma data_seg(pop)

static void
_get_programs(_Outptr_result_buffer_(*count) program_entry_t** programs, _Out_ size_t* count)
{
    *programs = _programs;
    *count = 1;
}

static void
_get_version(_Out_ bpf2c_version_t* version)
{
    version->major = 0;
    version->minor = 11;
    version->revision = 0;
}

static void
_get_map_initial_values(_Outptr_result_buffer_(*count) map_initial_values_t** map_initial_values, _Out_ size_t* count)
{
    *map_initial_values = NULL;
    *count = 0;
}

metadata_table_t divide_by_zero_metadata_table = {
    sizeof(metadata_table_t), _get_programs, _get_maps, _get_hash, _get_version, _get_map_initial_values};
