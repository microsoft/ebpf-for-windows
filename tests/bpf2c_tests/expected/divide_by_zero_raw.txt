// Copyright (c) Microsoft Corporation
// SPDX-License-Identifier: MIT

// Do not alter this generated file.
// This file was generated from divide_by_zero.o

#include "bpf2c.h"

static map_entry_t _maps[] = {
{ NULL, { 2, 4, 4, 1, 0, 0, 0, 0,  }, "test_map" },
};

static void _get_maps(_Outptr_result_buffer_(*count) map_entry_t** maps, _Out_ size_t* count)
{
	*maps = _maps;
	*count = 1;
}

static helper_function_entry_t divide_by_zero_helpers[] = {
{ NULL, 1, "helper_id_1"},
};

static GUID divide_by_zero_program_type_guid = {0xf1832a85, 0x85d5, 0x45b0, {0x98, 0xa0, 0x70, 0x69, 0xd6, 0x30, 0x13, 0xb0}};
static GUID divide_by_zero_attach_type_guid = {0x85e0d8ef, 0x579e, 0x4931, {0xb0, 0x72, 0x8e, 0xe2, 0x26, 0xbb, 0x2e, 0x9d}};
static uint16_t divide_by_zero_maps[] = {
0,
};

static uint64_t divide_by_zero(void* context)
{
#line 27 "sample/divide_by_zero.c"
	// Prologue
#line 27 "sample/divide_by_zero.c"
	uint64_t stack[(UBPF_STACK_SIZE + 7) / 8];
#line 27 "sample/divide_by_zero.c"
	register uint64_t r0 = 0;
#line 27 "sample/divide_by_zero.c"
	register uint64_t r1 = 0;
#line 27 "sample/divide_by_zero.c"
	register uint64_t r2 = 0;
#line 27 "sample/divide_by_zero.c"
	register uint64_t r3 = 0;
#line 27 "sample/divide_by_zero.c"
	register uint64_t r4 = 0;
#line 27 "sample/divide_by_zero.c"
	register uint64_t r5 = 0;
#line 27 "sample/divide_by_zero.c"
	register uint64_t r6 = 0;
#line 27 "sample/divide_by_zero.c"
	register uint64_t r10 = 0;

#line 27 "sample/divide_by_zero.c"
	r1 = (uintptr_t)context;
#line 27 "sample/divide_by_zero.c"
	r10 = (uintptr_t)((uint8_t*)stack + sizeof(stack));

	// EBPF_OP_MOV64_IMM pc=0 dst=r6 src=r0 offset=0 imm=0
#line 27 "sample/divide_by_zero.c"
	r6 = IMMEDIATE(0);
	// EBPF_OP_STXW pc=1 dst=r10 src=r6 offset=-4 imm=0
#line 29 "sample/divide_by_zero.c"
	*(uint32_t *)(uintptr_t)(r10 + OFFSET(-4)) = (uint32_t)r6;
	// EBPF_OP_MOV64_REG pc=2 dst=r2 src=r10 offset=0 imm=0
#line 29 "sample/divide_by_zero.c"
	r2 = r10;
	// EBPF_OP_ADD64_IMM pc=3 dst=r2 src=r0 offset=0 imm=-4
#line 29 "sample/divide_by_zero.c"
	r2 += IMMEDIATE(-4);
	// EBPF_OP_LDDW pc=4 dst=r1 src=r0 offset=0 imm=0
#line 30 "sample/divide_by_zero.c"
	r1 = POINTER(_maps[0].address);
	// EBPF_OP_CALL pc=6 dst=r0 src=r0 offset=0 imm=1
#line 30 "sample/divide_by_zero.c"
	r0 = divide_by_zero_helpers[0].address
#line 30 "sample/divide_by_zero.c"
	(r1, r2, r3, r4, r5);
#line 30 "sample/divide_by_zero.c"
	if ((divide_by_zero_helpers[0].tail_call) && (r0 == 0)) return 0;
	// EBPF_OP_JEQ_IMM pc=7 dst=r0 src=r0 offset=3 imm=0
#line 31 "sample/divide_by_zero.c"
	if (r0 == IMMEDIATE(0)) goto label_1;
	// EBPF_OP_LDXW pc=8 dst=r1 src=r0 offset=0 imm=0
#line 32 "sample/divide_by_zero.c"
	r1 = *(uint32_t *)(uintptr_t)(r0 + OFFSET(0));
	// EBPF_OP_MOV64_IMM pc=9 dst=r6 src=r0 offset=0 imm=100000
#line 32 "sample/divide_by_zero.c"
	r6 = IMMEDIATE(100000);
	// EBPF_OP_DIV64_REG pc=10 dst=r6 src=r1 offset=0 imm=0
#line 32 "sample/divide_by_zero.c"
	if (r1 == 0) { division_by_zero(10); return 0; }
#line 32 "sample/divide_by_zero.c"
	r6 /= r1;
label_1:
	// EBPF_OP_MOV64_REG pc=11 dst=r0 src=r6 offset=0 imm=0
#line 35 "sample/divide_by_zero.c"
	r0 = r6;
	// EBPF_OP_EXIT pc=12 dst=r0 src=r0 offset=0 imm=0
#line 35 "sample/divide_by_zero.c"
	return r0;
#line 35 "sample/divide_by_zero.c"
}
#line __LINE__ __FILE__

static program_entry_t _programs[] = {
	{ divide_by_zero, "xdp", "divide_by_zero", divide_by_zero_maps, 1, divide_by_zero_helpers, 1, 13, &divide_by_zero_program_type_guid, &divide_by_zero_attach_type_guid, },
};

static void _get_programs(_Outptr_result_buffer_(*count) program_entry_t** programs, _Out_ size_t* count)
{
	*programs = _programs;
	*count = 1;
}


metadata_table_t divide_by_zero_metadata_table = { _get_programs, _get_maps };
